src/Halogen.purs

```diff
@@ -15,19 +15,22 @@ module Halogen
 import Prelude
 
 import Control.Coroutine as CR
-
 import Data.Lazy (defer)
-
+import Data.Maybe (Maybe)
 import Halogen.Component (Component, ComponentSpec, ComponentSlot, ComponentSlotSpec, component, mkComponent, hoist, componentSlot, unComponent, unComponentSlot)
 import Halogen.Data.Slot (Slot)
 import Halogen.HTML (ComponentHTML, ComponentHTML')
 import Halogen.HTML.Core (AttrName(..), ClassName(..), Namespace(..), PropName(..), ElemName(..))
 import Halogen.Query (Action, HalogenF(..), HalogenM, HalogenM'(..), HalogenQ(..), RefLabel(..), Request, SubscriptionId, ForkId, action, fork, kill, get, getHTMLElementRef, getRef, gets, lift, liftAff, liftEffect, modify, modify_, put, query, queryAll, raise, request, subscribe, subscribe', unsubscribe)
 
 -- | A record produced when the root component in a Halogen UI has been run.
--- | `query` allows external sources to query the root component and `subscribe`
--- | allows external consumers to receive messages raised by the root component.
+-- |
+-- | - `query` allows external sources to query the root component
+-- | - `subscribe` allows external consumers to receive messages raised by the
+-- |   root component
+-- | - `dispose` stops running the UI and finalizes the root component
 type HalogenIO f o m =
-  { query :: f ~> m
+  { query :: forall a. f a -> m (Maybe a)
```

Comment:
I figured since `query` is normally partial, it wasn't any great pain to have this one partial either. Plus it means if we ever do integrate the `PartialF` style stuff we talked about in the other issue at least we're already set up for it here again.

I know the exception channel in `Aff` is there for a reason, but I much prefer to avoid it, since it's very easy to forget that you need to `try` anywhere that might fail, since the types don't communicate failure is "likely".

Reply:
Makes sense 👍